### PR TITLE
feat: Delete a user

### PR DESCRIPTION
**Pull request by idowuseyi**
**What does the PR do?**

This PR is about deleting a user

**How should this be manually tested?**

1. Login to a user account, and get the access token
2. Send a DELETE request to `/api/v1/users/:user_id`

Successful Response

On a successful user delete, the API should return a 200 SUCCESS status code.

```json
{
      "status": "success",
      "message": "Account Deleted Successfully",
      "data": {
            "user_id": "a9f4b3b0-4b3b-4b3b-4b3b-4b3b4b3b4b3b",
            "name": "John Doe",
            "phone_number":"+23467891234",
      }
}
```

Authentication Error:

When the current user does not have permission:

```json
{
          "status": "error",
          "message": "Bad Request",
          "status_code": "HttpStatus.BAD_REQUEST",
 }
```

Invalid UserID Error

If a wrong or invalid user_id is used:

```json
{
  "status": "Error",
  "message": "user_id is required",
  "status_code": 400
}
```

 **Checklist of what you have just done**
- [X ] Define API endpoint to soft delete a user.
- [X ]  Validate and handle authentication token.
- [ X]  Implement logic to soft delete user information from the database.
- [ X]  Handle success and error responses.
- [ X]  Test Successful response

**Test Screenshot**

Screenshot from 2024-07-24 23-47-26

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/167)

### CODE REVIEW

**Comment by idowuseyi** [comment]
Ttile updated

File: src/modules/user/user.service.ts [hunk 1]
```diff
@@ -25,6 +18,24 @@ export default class UserService {
...
   ) {}
 
+  async findOne(user_id: string): Promise<User> {
+    const user = await this.userRepository.findOne({ where: { id: user_id } });
+    if (!user) {
+      throw new NotFoundException(`User with id ${user_id} not found`);
+    }
+    return user;
+  }
+
```
Comment (davidolanrewaju) [inline src/modules/user/user.service.ts:28]: Wrap your db call in a try catch block for easier debugging

File: src/modules/user/user.service.ts [hunk 2]
```diff
@@ -25,6 +18,24 @@ export default class UserService {
...
+  async softDelete(user_id: string): Promise<User> {
```
Comment (davidolanrewaju) [inline src/modules/user/user.service.ts:29]: Wrap your db call in a try catch block

Comment (davidolanrewaju) [inline src/modules/user/user.service.ts:29]: Wrap your db call in a try catch block for easier debugging

File: src/modules/user/user.service.ts [hunk 3]
```diff
@@ -25,6 +18,24 @@ export default class UserService {
...
+    const user = await this.findOne(user_id);
```
Comment (davidolanrewaju) [inline src/modules/user/user.service.ts:30]: Wrap your db call in a try catch block for easier debugging

File: src/modules/user/user.service.ts [hunk 4]
```diff
@@ -25,6 +18,24 @@ export default class UserService {
     private userRepository: Repository<User>
   ) {}
 
+  async findOne(user_id: string): Promise<User> {
+    const user = await this.userRepository.findOne({ where: { id: user_id } });
```
Comment (davidolanrewaju) [inline src/modules/user/user.service.ts:22]: Wrap your db call in a try catch block for easier debugging

Comment (idowuseyi) [reply src/modules/user/user.service.ts]: Done!

**Review by Heba-WebDev — APPROVED** [review]

File: src/modules/user/user.service.ts (hunk 3 above)
Comment (idowuseyi) [reply src/modules/user/user.service.ts:41]: Done!

File: src/modules/user/user.service.ts (hunk 4 above)
Comment (idowuseyi) [reply src/modules/user/user.service.ts]: Done!

**Review by davidolanrewaju — APPROVED** [review]

**Comment by Heba-WebDev** [comment]
@idowuseyi your new changes are failing the CI/CI pipeline so look at this please

**Comment by idowuseyi** [comment]
Fixed the test failure already.

**Comment by idowuseyi** [comment]
Merge conflict resolved.

**Review by davidolanrewaju — APPROVED** [review]

**Comment by idowuseyi** [comment]
Fixed merge conflict.

**Pull request closed without merging**